File: jar-common/src/main/java/com/beancounter/common/identity/CallerRef.java
package com.beancounter.common.identity;

import com.beancounter.common.utils.DateUtils;
import com.beancounter.common.utils.KeyGenUtils;
import java.io.Serializable;
import java.util.UUID;
import javax.persistence.Embeddable;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * Uniquely identifies a transaction, within a batch from a data provider.
 *
 * @author mikeh
 * @since 2019-02-10
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Embeddable
public class CallerRef implements Serializable {
  private String provider;
  private String batch;
  private String callerId;

  // Fill in any missing default values from the supplied Id
  public static CallerRef from(CallerRef id) {
    CallerRef result = CallerRef.builder()
        .build();

    result.setProvider(id == null || id.getProvider() == null ? "BC" : id.provider);
    result.setBatch(id == null || id.getBatch() == null ? new DateUtils().today() : id.batch);
    result.setCallerId(id == null || id.getCallerId() == null
        ? KeyGenUtils.format(UUID.randomUUID()) : id.callerId);
    return result;
  }
}

File: jar-common/src/main/kotlin/com/beancounter/common/model/CallerRef.kt
package com.beancounter.common.model

import com.beancounter.common.utils.DateUtils
import com.beancounter.common.utils.KeyGenUtils
import jakarta.persistence.Embeddable

/**
 * Uniquely identifies a transaction, within a batch from a data provider.
 *
 * @author mikeh
 * @since 2019-02-10
 */
@Embeddable
data class CallerRef(
    // The System that owns the transaction
    var provider: String = "",
    var batch: String = "",
    var callerId: String = "",
) {
    /**
     * Helper methods to deal with CallerRef Objects
     */
    companion object {
        private const val DEFAULT_PROVIDER = "BC"
        private val dateUtils = DateUtils()
        private val keyGenUtils = KeyGenUtils()

        @JvmStatic
        fun from(callerRef: CallerRef): CallerRef {
            val provider = callerRef.provider.ifBlank { DEFAULT_PROVIDER }
            val batch =
                callerRef.batch.ifBlank {
                    dateUtils.getFormattedDate().toString().replace(
                        "-",
                        "",
                    )
                }
            val callerId = callerRef.callerId.ifBlank { keyGenUtils.id }
            return CallerRef(
                provider,
                batch,
                callerId,
            )
        }
    }
}
